### ISSUE TITLE
partial matching with dollar sign

### ISSUE THREAD

**Issue by kellijohnson** [ss3sim/ss3sim#188]
I am receiving the following error message after running a scenario with just length and age comps, where before I do not remember seeing this error.
```46: In `$.data.frame`(agecomp, Flt) : Name partially matched in data frame
47: In `$.data.frame`(calcomp, Flt) : Name partially matched in data frame
48: In `$.data.frame`(calcomp, Flt) : Name partially matched in data frame
49: In `$.data.frame`(a, Flt) : Name partially matched in data frame
50: In `$.data.frame`(a, Flt) : Name partially matched in data frame``
Is this something internal to the computer that I am running it on, although I tried setting ``options(warnPartialMatchDollar = TRUE)`` in my R session and I still got the error. I know we are not supposed to do partial matching for argument names in functions but is the partial matching within the datfile dataframe bad as well?

**Comment by seananderson** [comment]
Not sure what's going on here. I guess somewhere there's something like this?

```
d <- data.frame(this = 1:3, that = 2:4)
d$thi
## [1] 1 2 3
```

Probably not ideal. CRAN checks may never notice and complain since they don't exercise the full `run_ss3sim` function.

Any chance of passing on the code that caused this or hoping there's still some traceback info after?

There's also the possibility of setting `options(error = recover)` and `options(warn = 2)` to pop you into debugging mode with a warning.

Given it all involves `Flt`, could it be `FltSvy` and `Flt`?

**Comment by colemonnahan** [comment]
This is almost certainly in the `clean_data` function. I recognize my bad code! @kellijohnson recently changed how we subset the data frames since `r4ss` uses different column names than the `data.ss_new` file, so she hard coded it by col number. I would guess this raised a new issue?

@kellijohnson do you mind looking into this?

**Comment by kellijohnson** [comment]
I actually got @taylori to change r4ss and so the names are the same. 
Thus I ended up using your tactic of partial matching. I am fine with 
fixing it, but I think we should have a bit of a brain storm on the best 
way forward. Should we use grep with column names to find something that 
is close to "Fl" and then hard code for that column name (basically do 
our own partial match), hard code for the third column, or find a way to 
turn off the warnings for the partial match? I am unclear on what is the 
best way forward for development purposes.

**Comment by taylori** [comment]
Having suggested the not great partial matching solution, I would now suggest removing it (though I'm not sure why the error is occurring).

What I would consider doing is something like

    if("FltSvy" %in% names(MeanSize_at_Age_obs)){
      ... MeanSize_at_Age_obs$FltSvy...
    }
    if("Fleet" %in% names(MeanSize_at_Age_obs)){
      ... MeanSize_at_Age_obs$FltSvy...
    }

Also, I've been confusing myself editing both the r4ss master and a branch called "TA1.8testing". I'm hoping to merge these at some point in the weeks ahead, but it may be a while. Could somebody let me know if changes for ss3sim need to be in one place or the other or both. It looks like the change for @kellijohnson was just in the branch.

**Comment by taylori** [comment]
Code below is no longer identical in both places, just in case anybody was confused.

    if("FltSvy" %in% names(MeanSize_at_Age_obs)){
      ... MeanSize_at_Age_obs$FltSvy...
    }
    if("Fleet" %in% names(MeanSize_at_Age_obs)){
      ... MeanSize_at_Age_obs$Fleet...
    }

**Comment by seananderson** [comment]
I agree that some kind of explicit matching like what @taylori suggested is probably the best way forward. Clear to read and unlikely to create strange bugs down the road. If it affects a lot of code then maybe best to do our own grep and then alter column names once.

As for branches on r4ss, I'm sure for a little while we can work around whatever branch we need to use. In general, it's probably a good goal for the GitHub master branches of r4ss and ss3sim to work with each other (likewise for the CRAN versions to be in sync).

**Comment by taylori** [comment]
One additional thought.
Rick Methot has been working on SS version 3.3 for a while, and one of the
changes that he's making is breaking down the divisions between fleets and
surveys. This will make it much easier to add or subtract fishing fleets
(which previously required renumbering the index of all the surveys). I
believe that he's going to be using "fleet" throughout the code instead of
"FltSvy".

Hopefully that would reduce some of the inconsistencies that are found in
SS and mirrored (imperfectly) in r4ss which are tripping up ss3sim.
-Ian

**Comment by colemonnahan** [comment]
So we don't know why the warning is arising? I don't have a strong preference except that it should work ASAP so we can move forward. When SS3.3 comes out we'll have to do a round of upgrades (the unit tests will come in VERY handy then) so we could push it back until then, if we do end up supporting it...

**Comment by seananderson** [comment]
I was in the process of streamlining the unit tests when I realized that the test for `sample_mlacomp` was failing. Looks like `change_data` (via all the `make_dummy_dat_...` functions) is creating the column name `Flt` but many of the sampling functions are matching based on `FltSvy`.

Can I switch all of one or the other to be the same or is this going to break other things?

Example failing code:

```r
# load_all() or library("ss3sim")
datfile <- r4ss::SS_readdat(system.file("extdata/models/cod-om/codOM.dat", package = "ss3sim"),
  verbose = FALSE)
fctl <- system.file("extdata/models/cod-om/codOM.ctl", package = "ss3sim")
newdat <- change_data(datfile, outfile = "codOM-temp.dat",
types = c("age", "mla"), fleets = 1, years = 2000:2012, write_file = FALSE)
set.seed(123)
out <- sample_mlacomp(newdat, outfile = "ignore.dat", fleets = 1, ctlfile = fctl,
  Nsamp = list(rep(50, 13)), years = list(2000:2012), write_file = FALSE,
  mean_outfile = NULL)
```

**Comment by kellijohnson** [comment]
The sample functions use an augmented datfile 
```r
rdatfile <- r4ss::SS_readdat(system.file("extdata/models/cod-om/codOM.dat", 
  package = "ss3sim"), verbose = FALSE)  
datfile <- change_fltname(datfile)  
fctl <- system.file("extdata/models/cod-om/codOM.ctl", package = "ss3sim")
newdat <- change_data(datfile, outfile = "codOM-temp.dat", 
  types = c("age", "mla"), fleets = 1, years = 2000:2012, write_file = FALSE)
set.seed(123)
newdat <- change_fltname(newdat)
out <- sample_mlacomp(newdat, outfile = "ignore.dat", fleets = 1, ctlfile = fctl, 
  Nsamp = list(rep(50, 13)), years = list(2000:2012), write_file = FALSE, mean_outfile = NULL)
```

2a7027d8cb0d7e99a867a048d00d049e17b39ca9

1effade70c2f1bc37493e1fb5c21833c0c39ceed is needed to add "Flt" to the search, but as mentioned in the commit message this may be better fixed within the sampling functions rather than changing the data fed to the sampling functions.

**Comment by seananderson** [comment]
Thanks Kelli. OK, so it looks like this not a problem within `ss3sim_base`, only within the unit tests where we haven't added a call to `change_fltname` yet (I'll fix that). I guess I'm fine with that solution for now. I see the problem is that `change_data` isn't always called right before a sampling function so we can't rely on that output. Sometimes we're reading data files in after an OM run.

I was going to re-arrange the unit tests and add in tests that automatically test `run_ss3sim` across a range of scenarios but aren't tested on CRAN.

**Comment by kellijohnson** [comment]
I did a search for SS_readdat and thought I placed change_fltname before 
it everywhere but inside the change_e function (which the calls 
SS_readdat is now removed in the branch externalgrowth which I will do a 
pull request soon).

If we are going to stick with change_fltname it should be added to the 
examples of each sampling function where SS_readdat is used. Let me know 
if I should do this.
